### PR TITLE
[feature] Delete node in API

### PR DESCRIPTION
**Pull request by pattisdr**
<h1> Purpose </h1>
Issue #3143.  Adding capabilities to delete node via API.   
<ul>
<li> User sends DELETE request containing node id</li>
<li> Server returns error, asking the user to confirm they want to delete project, along with a new url </li>
<li> User sends DELETE request to second url, containing node id and token </li>
<li> Provided token and node id are valid, node is deleted </li>
</ul>

<h1> Changes </h1>
<ul> 
<li> token creator added to utils.py, which is part of the new url returned </li>
<li> url added to urls.py </li>
<li> translatable warning text added to language.py </li>
<li> NodeDetail expanded to include first delete node logic and new class NodeDeleteConfirm added for second delete node logic in views.py </li>
<li> tests for deleting node in test_views.py </li>
<li> correction to test_deletes_private_node_pointer_logged_in_non_contributor.  URL and assert were incorrect. (I originally wrote this test). </li>
</ul> 
<h1> Side Effects </h1>
Two lines added in Swagger documentation 
screen shot 2015-06-16 at 2 28 01 pm

### CODE REVIEW

File: api/base/utils.py [hunk 1]
```diff
@@ -58,3 +59,9 @@ def waterbutler_url_for(request_type, provider, path, node_id, token, obj_args=N
 
     url.args.update(query)
     return url.url
+
+def node_token_creator(nodeid, userid):
```
Comment (lyndsysimon) [inline api/base/utils.py]: Please add at least a single-line docstring.

ref: http://cosdev.readthedocs.org/en/latest/style_guides/python.html#documentation

Comment (lyndsysimon) [inline api/base/utils.py]: Parameters (and all variables) should use snake case. `nodeid` should be `node_id`.

ref: http://cosdev.readthedocs.org/en/latest/style_guides/python.html#naming

File: api/nodes/urls.py [hunk 2]
```diff
@@ -7,6 +7,7 @@
     # url(r'^blog/', include('blog.urls')),
     url(r'^$', views.NodeList.as_view(), name='node-list'),
     url(r'^(?P<pk>\w+)/$', views.NodeDetail.as_view(), name='node-detail'),
+    url(r'^(?P<pk>\w+)/confirm/(?P<token>\w+)$', views.NodeDeleteConfirm.as_view(), name='node-delete-confirm'),
```
Comment (lyndsysimon) [inline api/nodes/urls.py]: The majority of rules in this file end with a trailing slash. Is there any reason this one shouldn't?

File: api/nodes/views.py [hunk 3]
```diff
@@ -11,7 +11,10 @@
 from .serializers import NodeSerializer, NodePointersSerializer, NodeFilesSerializer
 from api.users.serializers import ContributorSerializer
 from .permissions import ContributorOrPublic, ReadOnlyIfRegistration, ContributorOrPublicForPointers
-
+from api.base.utils import node_token_creator, absolute_reverse
+from website.language import BEFORE_DELETE
+from rest_framework import serializers
+from django.utils.translation import ugettext_lazy as _
```
Comment (lyndsysimon) [inline api/nodes/views.py]: Imports from a third-party library, in this case `rest_framework` and `django`, should be in the second import section.

ref: http://cosdev.readthedocs.org/en/latest/style_guides/python.html#imports

I personally prefer that imports be in alphabetical order within those sections, but that's not part of the styleguide, and not something I'd ask you to change. It's just my OCD nature showing through :)

File: api/nodes/views.py [hunk 4]
```diff
@@ -27,7 +30,6 @@ def get_node(self):
         self.check_object_permissions(self.request, obj)
         return obj
 
-
```
Comment (lyndsysimon) [inline api/nodes/views.py]: Class definitions should be surrounded by two blank lines. This happens a few others places in this file as well.

ref: https://www.python.org/dev/peps/pep-0008/#blank-lines

File: website/language.py [hunk 5]
```diff
@@ -2,7 +2,7 @@
 # -*- coding: utf-8 -*-
 """Various text used throughout the website, e.g. status messages, errors, etc.
 """
-
+from django.utils.translation import ugettext_lazy as _
```
Comment (lyndsysimon) [inline website/language.py]: I think this is a *great* idea - but I'm uncomfortable with a django method being imported into this file. Internationalization isn't something that we've yet addressed site-wide on the OSF, either. Maybe we're considering it from the start with the API, but if so I'm not aware of that decision having been made.

@brianjgeiger, could you weigh in on this? 

Here's the Django docs on this method: https://docs.djangoproject.com/en/1.8/topics/i18n/translation/#standard-translation

Comment (brianjgeiger) [inline website/language.py]: This was a request from me. We can move the variable back to the API side, I'm fine with that. There was some evolution here, and this might have evolved the wrong way. 

@pattisdr, let's go ahead and move the text into a place in the API that is visible for whatever needs to see this.

Comment (lyndsysimon) [inline website/language.py]: After speaking with BrianG, let's move this `/api/language.py`. That file doesn't exist, so you'll need to create it.

File: api/nodes/urls.py (hunk 2 above)
Comment (lyndsysimon) [inline api/nodes/urls.py]: After speaking with BrianG, the inconsistency shouldn't be there - we should be terminating URLs with a slash. If there is a compelling reason to deviate from that standard we can re-evaluate at a late date.

As for the existing route that is already in place without a trailing slash, let's leave that be in this PR, and we'll address it separately.

File: api/nodes/views.py (hunk 4 above)
Comment (sloria) [inline api/nodes/views.py]: FWIW, this rule is explicitly ignored by our setup.cfg: https://github.com/CenterForOpenScience/osf.io/blob/05d7f0e135a6bd5bb8ca196646bfbb85acf2dc25/setup.cfg#L6.

I find the extra whitespace to be unnecessary; it does not significantly affect readability.

File: tests/api_tests/nodes/test_views.py [hunk 6]
```diff
@@ -605,6 +605,81 @@ def test_partial_update_private_project_logged_in_non_contributor(self):
...
+        # a little better
+        assert_equal(res.status_code, 403)
+
+    def test_deletes_public_node_logged_in(self):
+        res = self.app.delete(self.public_url, auth=self.basic_auth_two, expect_errors=True)
+        assert_equal(res.status_code, 403)
+        assert_equal(self.public_project.is_deleted, False)
+
+        res = self.app.delete(self.public_url, auth=self.basic_auth, expect_errors=True)
+        assert_equal(res.status_code, 400)
```
Comment (lyndsysimon) [inline tests/api_tests/nodes/test_views.py]: This route returning a 400 (or any 4xx) feels wrong to me. My understanding of the 4xx series codes is that they should be sent only when there is an error made by the client. As this is the intended flow for a node deletion, returning a client error doesn't make sense.

After careful review of the applicable RFCs, I propose we use [`202 ACCEPTED`](https://tools.ietf.org/html/rfc7231#section-6.3.3) instead. 

> The 202 (Accepted) status code indicates that the request has been
   accepted for processing, but the processing has not been completed.
   The request might or might not eventually be acted upon, as it might
   be disallowed when processing actually takes place.  There is no
   facility in HTTP for re-sending a status code from an asynchronous
   operation.

> The 202 response is intentionally noncommittal.  Its purpose is to
   allow a server to accept a request for some other process (perhaps a
   batch-oriented process that is only run once per day) without
   requiring that the user agent's connection to the server persist
   until the process is completed.  The representation sent with this
   response ought to describe the request's current status and point to
   (or embed) a status monitor that can provide the user with an
   estimate of when the request will be fulfilled.

By returning a 202, we're effectively saying "OK, cool. You want that node deleted, and I hear you loud and clear. I'll take care of it, as soon as I'm ready."

Since the spec also allows for a status monitor to be returned. The mechanism for that is undefined, but I assume it would be either a `Location: ` header, or in the body of the response. Our API could then specify that this URL must be hit with a `DELETE` request before the node will actually be deleted.

This seems to satisfy the desire for a two-step process for deleting a node, while avoiding a response indicating an error.

**Comment by lyndsysimon** [comment]
First pass complete. :computer:

File: tests/api_tests/nodes/test_views.py (hunk 6 above)
Comment (brianjgeiger) [inline tests/api_tests/nodes/test_views.py]: Awesome, thanks.

**Comment by pattisdr** [comment]
Addressed comments in first pass.

File: api/nodes/views.py [hunk 7]
```diff
@@ -11,6 +11,11 @@
 from .serializers import NodeSerializer, NodePointersSerializer, NodeFilesSerializer
 from api.users.serializers import ContributorSerializer
 from .permissions import ContributorOrPublic, ReadOnlyIfRegistration, ContributorOrPublicForPointers
+from django.utils.translation import ugettext_lazy as _
```
Comment (lyndsysimon) [inline api/nodes/views.py]: This should be moved to the block containing `from modularodm import Q`, as it's a third-party library.

File: api/nodes/views.py [hunk 8]
```diff
@@ -92,17 +97,56 @@ class NodeDetail(generics.RetrieveUpdateAPIView, NodeMixin):
...
         return {'request': self.request}
 
+    # overrides RetrieveUpdateDestroyAPIView
+    def perform_destroy(self, instance):
+        user = self.request.user
+        node = self.get_node()
+        token = node_token_creator(node._id, user._id)
+        url = absolute_reverse('nodes:node-delete-confirm', kwargs={'node_id': node._id, 'token': token})
+        delete_warning = BEFORE_DELETE_NODE.format(_(node.title))
+        raise Accepted([delete_warning, url])
```
Comment (lyndsysimon) [inline api/nodes/views.py]: It feels very weird to me to raise an exception in the default case. Could we not return a response with the correct status code, without raising it as an exception?

File: tests/api_tests/nodes/test_views.py [hunk 9]
```diff
@@ -605,6 +610,85 @@ def test_partial_update_private_project_logged_in_non_contributor(self):
...
+
+    def test_deletes_public_node_logged_in(self):
+        res = self.app.delete(self.public_url, auth=self.basic_auth_two, expect_errors=True)
+        assert_equal(res.status_code, 403)
+        assert_equal(self.public_project.is_deleted, False)
+
+        res = self.app.delete(self.public_url, auth=self.basic_auth, expect_errors=True)
+        assert_equal(res.status_code, 202)
+        assert_equal(self.public_project.is_deleted, False)
+        detail = ast.literal_eval(res.json['detail'])
```
Comment (lyndsysimon) [inline tests/api_tests/nodes/test_views.py]: eval == evil :)

It's not a huge problem in a test, but I'm not quite understanding why it's needed here. If you've got a string that evals into a list, couldn't you just call `json.loads` on it?

**Comment by lyndsysimon** [comment]
Looks great overall! First pass review complete.

File: api/nodes/views.py (hunk 8 above)
Comment (brianjgeiger) [inline api/nodes/views.py]: Yeah, it's weird. But it also stops DRF from deleting the node at the end. It's functioning as an exception from our end, but it's reporting to the user as success, which is accurate but weird. There may be a better way to do it.

Comment (lyndsysimon) [inline api/nodes/views.py]: Ahh, that makes sense then. Carry on :)

File: tests/api_tests/nodes/test_views.py (hunk 9 above)
Comment (pattisdr) [inline tests/api_tests/nodes/test_views.py]: A unicode object is being returned.  There's a list inside. First index is warning message, second is url.  I want the url.   

 `u"[u'Are you sure you want to delete The meaning of life? It will no longer be available to other contributors on the project. Submit to new url to continue.', 'http://localhost:8000/v2/nodes/txbp9/confirm/a8bc9efd1d0be3abffd810528f49a7ec/']" `

If I try to access the url by doing `res.json['detail'][1]`, I get u'u'.  If I try `json.loads`, I get ValueError: No JSON object could be decoded.  The only way I know how to get what I want is eval[res.json['detail'])[1].  I know eval is bad...what else should I try?

Comment (brianjgeiger) [inline tests/api_tests/nodes/test_views.py]: What are you using json.loads on when it fails?

Comment (pattisdr) [inline tests/api_tests/nodes/test_views.py]: If I do `json.loads` on the unicode object above, it returns ValueError: No JSON object could be decoded.  

If I do `json.loads(res.json['detail'])`, I get a ValueError: Expecting property name enclosed in double quotes.

Comment (brianjgeiger) [inline tests/api_tests/nodes/test_views.py]: so specifically, json.loads(res.json) doesn't work?

Comment (pattisdr) [inline tests/api_tests/nodes/test_views.py]: Yes, I haven't gotten it to work.

Comment (brianjgeiger) [inline tests/api_tests/nodes/test_views.py]: So all that being said, now that I'm looking at the response object, this should be a dictionary, not a list. The URL should be a… relation link, I believe. We might want to have a serializer specifically for this response so people will know the structure of what they're getting back before they get it. Probably will have an id, a warning message, and the link dictionary with the relation field.

Comment (brianjgeiger) [inline tests/api_tests/nodes/test_views.py]: I'm not positive a serializer makes sense for the response object, mind you, but it should definitely be a dictionary not a list.

Comment (pattisdr) [inline tests/api_tests/nodes/test_views.py]: Yes, I had tried a dictionary first, but ran into the same problems with accessing url, because it's a unicode object. 

My exception looked something like this: `raise Accepted({'warning': delete_warning, 'url':url})`  Then I couldn't access keys in the dictionary.  In the test, If I tried res.json['detail']['url'] it would tell me that string indices had to be integers.

Comment (brianjgeiger) [inline tests/api_tests/nodes/test_views.py]: Maybe put a breakpoint in at that line of your test and see if you can evaluate some expressions in that context to experiment.

Comment (pattisdr) [inline tests/api_tests/nodes/test_views.py]: Now a dictionary, but still retrieving url using eval.  Will try making a serializer and see what that looks like.

**Comment by pattisdr** [comment]
You guys were right.  I replaced `perform_destroy` in the view with `delete`.  Then I was able to raise a response with the 202 status code without using an exception and still prevent the actual delete.

Now the url can be easily retrieved from the response without eval.  Using an exception was causing the problems.

**Comment by brianjgeiger** [comment]
:tada:

**Comment by pattisdr** [comment]
Response to review complete

**Comment by lyndsysimon** [comment]
Woot!

It sounds like between the three of us we came up a better solution than any of us would have on our own :)

:dancers:

File: tests/api_tests/nodes/test_views.py (hunk 9 above)
Comment (lyndsysimon) [inline tests/api_tests/nodes/test_views.py]: `u''` shouldn't be a part of any string that's returned from Python - it basically means that `__repr__` is being called on something it shouldn't be.

One place where that happens by accident a lot is when you are expecting a string, get a dict, and explicitly convert it. For some reason, `dict.__str__` just calls `dict.__repr__`. If you ask me, that's a bug :)

Example:

```python
foo = {u'apple': u'pear'}
print(foo)
# {u'apple': u'pear'}

foo_coerced = str(foo)
print(repr(foo_coerced))
# u'{u\'apple\': u\'pear\'}'
```

I'm guessing that's what is happening hear - `str()` is getting called on an dict, when `json.dumps()` should be used instead.

**Comment by lyndsysimon** [comment]
I'm going to hold off marking this for Steve until we get one more pass in on that eval being used in the test. If it were only the use of eval, I wouldn't worry about it much - but I'm now thinking that the fact that it's required might be indicative of us testing for the wrong result.

File: tests/api_tests/nodes/test_views.py (hunk 9 above)
Comment (pattisdr) [inline tests/api_tests/nodes/test_views.py]: I'm not using `eval` in the test anymore, after I changed the exception to a `Response` object (containing the 202 status, a warning message, and the revised URL.)  To get the url in the test, I only need res.json['url'].

**Comment by lyndsysimon** [comment]
Yeah, I got confused with the history on this one. Nevermind :)

@sloria, this one is ready for you! :cake:

**Comment by brianjgeiger** [comment]
Trying this out, we should ensure the format is as close to a normal response as possible. We should wrap the data in dictionary called data. The url should be like the links inside node, so it should be links: { confirm_delete: { \<url\> } } and the node id should be one of the fields.

**Comment by brianjgeiger** [comment]
I thiiink the links object, in this case, should go in the data object. Links outside are for pagination, by and large.

**Comment by brianjgeiger** [comment]
Other than that, it looks great to me.

**Comment by pattisdr** [comment]
Responded to Brian's comments about links.

File: api/nodes/views.py [hunk 10]
```diff
@@ -92,17 +98,60 @@ class NodeDetail(generics.RetrieveUpdateAPIView, NodeMixin):
...
     def get_serializer_context(self):
         # Serializer needs the request in order to make an update to privacy
         return {'request': self.request}
 
+    def delete(self, request, node_id):
+        user = self.request.user
+        node = self.get_node()
+        token = node_token_creator(node._id, user._id)
+        url = absolute_reverse('nodes:node-delete-confirm', kwargs={'node_id': node._id, 'token': token})
+        delete_warning = BEFORE_DELETE_NODE.format(_(node.title))
```
Comment (sloria) [inline api/nodes/views.py]: There's no need to call `ugettext` on node titles, as that is a user input value.

File: api/nodes/views.py [hunk 11]
```diff
@@ -92,17 +98,60 @@ class NodeDetail(generics.RetrieveUpdateAPIView, NodeMixin):
...
+    permission_classes = (
+        ContributorOrPublic,
+        ReadOnlyIfRegistration,
+    )
+    serializer_class = NodeSerializer
+
+    def get_object(self):
+        return self.get_node()
+
+    def perform_destroy(self, instance):
```
Comment (sloria) [inline api/nodes/views.py]: Document method overrides with a line comment.

https://cosdev.readthedocs.org/en/latest/style_guides/python.html#method-overrides

File: api/nodes/views.py [hunk 12]
```diff
@@ -92,17 +98,60 @@ class NodeDetail(generics.RetrieveUpdateAPIView, NodeMixin):
...
+    """Projects and component details.
+
+    On the front end, nodes are considered 'projects' or 'components'. The difference between a project and a component
+    is that a project is the top-level node, and components are children of the project. There is also a category field
+    that includes the option of project. The categorization essentially determines which icon is displayed by the
+    Node in the front-end UI and helps with search organization. Top-level Nodes may have a category other than
+    project, and children nodes may have a category of project.
+    """
+    permission_classes = (
+        ContributorOrPublic,
```
Comment (sloria) [inline api/nodes/views.py]: `ContributorOrPublic` will not suffice; the user must have write permissions to the node; I'm not sure if there are permission classes to handle this (ping @brianjgeiger ), but the `Node#remove_node` will raise a `PermissionsError` if the user does not have write permissions. This error should be handled and reraised as the proper DRF error.

File: api/nodes/views.py [hunk 13]
```diff
@@ -92,17 +98,60 @@ class NodeDetail(generics.RetrieveUpdateAPIView, NodeMixin):
...
+        user = self.request.user
+        node = self.get_node()
+        token = node_token_creator(node._id, user._id)
+        url = absolute_reverse('nodes:node-delete-confirm', kwargs={'node_id': node._id, 'token': token})
+        delete_warning = BEFORE_DELETE_NODE.format(_(node.title))
+        return Response({'data': {'node_id': node._id, 'warning_message': delete_warning, 'links': {'confirm_delete': url}}}, status=status.HTTP_202_ACCEPTED)
+
+
+class NodeDeleteConfirm(generics.DestroyAPIView, NodeMixin):
+    """Projects and component details.
```
Comment (sloria) [inline api/nodes/views.py]: This docstring is inaccurate.

File: tests/api_tests/nodes/test_views.py [hunk 14]
```diff
@@ -605,6 +610,83 @@ def test_partial_update_private_project_logged_in_non_contributor(self):
...
+        res = self.app.delete(self.private_url, auth=self.basic_auth, expect_errors=True)
+        assert_equal(res.status_code, 202)
+        assert_equal(self.project.is_deleted, False)
+        returned_url = res.json['data']['links']['confirm_delete']
+
+        res = self.app.delete(returned_url, auth=self.basic_auth)
+        assert_equal(res.status_code, 204)
+        assert_equal(self.project.is_deleted, True)
+
+    def test_deletes_private_node_logged_in_non_contributor(self):
```
Comment (sloria) [inline tests/api_tests/nodes/test_views.py]: Please also add a test for a contributor that does not have proper permissions (i.e. a read-only contributor).

File: tests/api_tests/nodes/test_views.py [hunk 15]
```diff
@@ -580,6 +584,7 @@ def test_partial_update_public_project_logged_in(self):
         assert_equal(res.json['data']['title'], self.new_title)
         assert_equal(res.json['data']['description'], self.description)
         assert_equal(res.json['data']['category'], self.category)
+        print res.json
```
Comment (sloria) [inline tests/api_tests/nodes/test_views.py]: Remove print statement.

File: api/nodes/views.py [hunk 16]
```diff
@@ -92,17 +98,60 @@ class NodeDetail(generics.RetrieveUpdateAPIView, NodeMixin):
         ContributorOrPublic,
         ReadOnlyIfRegistration,
     )
-    serializer_class = NodeSerializer
 
-    # overrides RetrieveUpdateAPIView
```
Comment (sloria) [inline api/nodes/views.py]: This comment should be updated rather than deleted.

**Comment by sloria** [comment]
After some thought, I don't think a 2-step deletion process is necessary. Adding a confirmation step is more of client concern than something that should be implemented on the server. Sure, a confirmation step is a good idea *in a UI*, but why force every client to make 2 requests to make a deletion?

Thoughts, @brianjgeiger?

File: api/nodes/views.py (hunk 12 above)
Comment (brianjgeiger) [inline api/nodes/views.py]: If it's not admin-only, ContributorOrPublic should work, because it distinguishes the type of access method and responds appropriately.

```py
        if request.method in permissions.SAFE_METHODS:
            return obj.is_public or obj.can_view(auth)
        else:
            return obj.can_edit(auth)
```

Safe methods are read only. The rest require write permission.

**Comment by sloria** [comment]
Found a related SO thread: https://stackoverflow.com/questions/2539394/rest-http-delete-and-parameters -- consensus seems to be that confirmation is a UI concern, not a server concern.

**Comment by brianjgeiger** [comment]
The points at which I've added two-step are usually ones in which we want to return a specific message to warn people about consequences, especially if those consequences are unique to our workflow. It's not as big of a deal with deletion as it is with registrations, so I'm fine with making it one step. I would keep registration finalization as a two-step process, however.

File: tests/api_tests/nodes/test_views.py (hunk 14 above)
Comment (brianjgeiger) [inline tests/api_tests/nodes/test_views.py]: Ooh, we're probably going to need that test for the other write functionality. Not part of this PR, though.

**Comment by pattisdr** [comment]
Addressed comments, turned two-step delete process into one.

File: tests/api_tests/nodes/test_views.py [hunk 17]
```diff
@@ -605,6 +609,74 @@ def test_partial_update_private_project_logged_in_non_contributor(self):
...
+
+        self.public_project = ProjectFactory(is_public=True, creator=self.user)
+        self.public_url = '/{}nodes/{}/'.format(API_BASE, self.public_project._id)
+
+        self.fake_url = '/{}nodes/{}/'.format(API_BASE, '12345')
+
+    def test_deletes_public_node_logged_out(self):
+        res = self.app.delete(self.public_url, expect_errors=True)
+        # This is 403 instead of 401 because basic authentication is only for unit tests and, in order to keep from
+        # presenting a basic authentication dialog box in the front end. We may change this as we understand CAS
```
Comment (sloria) [inline tests/api_tests/nodes/test_views.py]: I'm not clear on why we're not returning 401 here; it seems to be the more appropriate status code.

Comment (sloria) [inline tests/api_tests/nodes/test_views.py]: ping @brianjgeiger

Comment (sloria) [inline tests/api_tests/nodes/test_views.py]: Is it just because we don't want to pop up the dialog in the browse-able API?

File: api/nodes/urls.py [hunk 18]
```diff
@@ -1,5 +1,5 @@
-from django.conf.urls import url
 from api.nodes import views
+from django.conf.urls import url
```
Comment (sloria) [inline api/nodes/urls.py]: 3rd party imports should go before application imports: https://cosdev.readthedocs.org/en/latest/style_guides/python.html#imports

**Comment by sloria** [comment]
Just a couple minor things to address and this is good to merge. :+1:

File: tests/api_tests/nodes/test_views.py (hunk 17 above)
Comment (brianjgeiger) [inline tests/api_tests/nodes/test_views.py]: That is exactly the reason.

Comment (brianjgeiger) [inline tests/api_tests/nodes/test_views.py]: Now that I've thought about it some more, do you think a custom error that also sends a 403 would be the way to go? Might be worth trying. It does bother me about the wrong status code.

Comment (sloria) [inline tests/api_tests/nodes/test_views.py]: I'm not sure that  custom error would be helpful. I'm also don't think that the browser basic auth behavior is a problem. The generic 401 response is what we want.

Comment (brianjgeiger) [inline tests/api_tests/nodes/test_views.py]: I don't want people just playing around with the API docs to send their password over basic auth. It's not a terribly secure mechanism.

Comment (brianjgeiger) [inline tests/api_tests/nodes/test_views.py]: But how about we fix this as a separate issue, since it exists in other places in the API already?

Comment (sloria) [inline tests/api_tests/nodes/test_views.py]: Sounds good.
